metaseq/data/streaming_src_tgt_dataset.py

```diff
@@ -164,3 +165,45 @@ def yield_src_tgt_blocks(iterable, block_size, drop_last, padding_idx):
             "src_block": src_block,
             "tgt_block": tgt_block,
         }
+
+
+def yield_src_tgt_single_sentences_pad_8(iterable, block_size, drop_last, padding_idx):
+    """Mimics sample-break-mode eos i.e. 1 example per sequence without any packing.
+    When multiple examples are packed into a single sequence, example tokens would attend
+    to tokens in neighbouring examples, which may be undesirable. This mode can
+    avoid that. Since there is no packing, this mode is considerably slower.
+    We round up the example length to a multiple of 8, pad to this length and
+    return the example as is, without packing, truncating to block_size in cases of
+    very long examples.
+    """
+
+    for idx, (src, tgt) in enumerate(iterable):
+        cur_src_block = []
+        cur_src_block_ids = []
+        cur_tgt_block = []
+        if src.numel() > block_size:
+            # truncate right side
+            # TODO: Enable left side truncation
+            src = src[:block_size]
+            tgt = tgt[:block_size]
+
+        cur_src_block.append(src)
+        cur_src_block_ids.append(idx)
+        cur_tgt_block.append(tgt)
+
+        # We round up to a multiple of 8 + 1, because later on
+        # one element is removed for src/target tensor creation
+        # which brings it back to a multiple of 8. block_size is
+        # already passed with + 1 included.
```

Comment:
Add an assertion for `block_size` being passed with + 1 included?

Reply:
Hard to know that here, block size is passed in streaming_language_modeling task with the +1.

Reply:
Is there a way to assert something around this?  This seems like it depends on very brittle logic...

Reply:
Technically, it will trigger the megatron assert - so it will not be something that will go unnoticed and cause issues. Not sure how to assert for this.